src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -809,8 +807,25 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
             // SPEC: better than all other function members, then the function member invocation is ambiguous and a binding-time 
             // SPEC: error occurs.
 
-            // UNDONE: This is a naive quadratic algorithm; there is a linear algorithm that works. Consider using it.
             var candidates = result.Results;
+            // Try to find a single best candidate
+            int bestIndex = GetTheBestCandidateIndex(left, right, candidates, ref useSiteDiagnostics);
+            if (bestIndex != -1)
+            {
+                // Mark all other candidates as worse
+                for (int i = 0; i < candidates.Count; ++i)
+                {
+                    if (i == bestIndex || candidates[i].Kind == OperatorAnalysisResultKind.Inapplicable)
+                    {
+                        continue;
+                    }
+
+                    candidates[i] = candidates[i].Worse();
```

Comment:
>candidates[i] = candidates[i].Worse(); [](start = 20, length = 38)

Consider inverting condition of the ```if``` and moving this statement inside it instead of using ```continue```.